go-controller/pkg/ovn/ovn.go

```diff
@@ -897,6 +897,19 @@ func (oc *Controller) syncNodeGateway(node *kapi.Node, hostSubnets []*net.IPNet)
 	if hostSubnets == nil {
 		hostSubnets, _ = util.ParseNodeHostSubnetAnnotation(node)
 	}
+
+	// Create the distributed gateway port that connects node switch to the cluster router.
+	// NOTE: We could have created the router port in ensureNodeLogicalNetwork() instead of here,
+	// but chassis ID is not available at that moment. We need the chassis ID to set the
+	// gateway-chassis, which in effect pins the logical switch to the current node in OVN.
+	// Otherwise, ovn-controller will flood-fill unrelated datapaths unnecessarily, causing scale
+	// problems.
+	if hostSubnets != nil {
+		if err := oc.syncNodeDGP(node.Name, l3GatewayConfig.ChassisID, hostSubnets); err != nil {
```

Comment:
Shouldn't this get moved under the `} else if hostSubnets != nil {` just below?

Reply:
Also, if we're going to use the chassisID outside of the gateway code we should just decouple that from the L3Gateway code on the node side. It's a separate annotation already, but handled by the gateway code. Should split that out.

Reply:
> Shouldn't this get moved under the `} else if hostSubnets != nil {` just below?

No, because this is regardless of the config.GatewayModeDisabled.

Reply:
> Also, if we're going to use the chassisID outside of the gateway code we should just decouple that from the L3Gateway code on the node side. It's a separate annotation already, but handled by the gateway code. Should split that out.

Agree. I am still getting familiar with the code, so I was trying to make as little change as possible. Let me see if I can change it that way with confidence.

Reply:
@dcbw I have decoupled the code from L3Gateway as suggested. Please take a look at the latest version.